### DISCUSSION TITLE
How do you set an initial Select menu option for a table's CreateAction?

### DISCUSSION THREAD

**Discussion by myregistration** [filamentphp/filament#13092] [Help]
### Package

Actions

### Package Version

v3

### How can we help you?

I've tried using mutateFormDataBeforeFill, beforeFill, afterFill in the relative `CreateRecord` and I've also tried `before`, `beforeFormFilled`, `afterFormFilled` in table's `CreateAction`.  I even tried hard-coding the `$data` array value for the select menu and created a test select menu, but nothing is working.  Are we supposed to set the `->default()` value in the select menu?  I would think mutating the `$data` array would set the initial select menu option.  If I `dd($data)` in the `beforeFormFilled` or `afterFormFilled` methods I can see the value is being set.  Thanks

Here's a test select menu I created in a `Filament Resource` class to test the relative lifecycle methods:

```
public static function form(Form $form): Form
{
	return $form
            ->schema([

                Select::make('category')
                    ->options([
                        'option1' => 'Option 1',
                        'option2' => 'Option 2',
                        'option3' => 'Option 3',
                    ])->default('option1')
                    ->required(),
```

```
public static function table(Table $table): Table
{
	return $table
	    ->headerActions([
		    Tables\Actions\CreateAction::make()
    		    ->beforeFormFilled(function (array $data, Component $livewire): array {

        		    $data['category'] = 'option2';

        		    return $data;
   		     }),
```

**Comment by leandrocfe** [comment]
Are you using CreateAction as a Modal?

**Reply by myregistration** [reply]
I've commented out the` getPages()` create url to use as a slideOver.  Should I convert this to an issue?  Do you have to use a select's `default()`?   I tried and can set the initial value using a closure in select menu's `->default()` method, but this is not ideal.  I would think you could just mutate the $data before it fills the form.

**Reply by danharrin** [reply]
Why are you using a CreateAction in the header of the table instead of the header of the page?

**Comment by photojo** [comment]
How did you solve the problem?

I do not want use ->default(). 

` public function mount()
    {
        $this->form->fill(['value'] => 'loremipsum');
    }`

Do not work in Panelbuilder. 

Please give me a hint. Thank you!

**Reply by myregistration** [reply]
Did you resolve your issue?  Are you using Filament Resource or it's Pages?  If so you can mutate the form $data array at different stages, like when it's loaded or before and after saved.  Unfortunately, they named the functions differently depending on if you are in a CreateAction within the Table actions array or List page, which is confusing.   I normally ask Laracasts AI to list the available methods, which is something it can usually get right ;).  

 It will probably be something like `mutateFormDataBeforeFill()`

```
    protected function mutateFormDataBeforeFill(array $data): array
    {
        $data['user_id'] = auth()->id();

        return $data;
    }
```

 or `mutateFormDataUsing()`

```
Tables\Actions\CreateAction::make()
                    ->mutateFormDataUsing(function (array $data, Component $livewire): array {

         $data['user_id'] = auth()->id();

         return $data;
}
```

Another way to modify field data is using` formatStateUsing() `        
          
`->formatStateUsing(fn ($state): string => $state ? Carbon::parse($state)->longAbsoluteDiffForHumans() . ' ago' : '')`

Let me know if this is the information you were looking for.

**Reply by photojo** [reply]
Hi Michelle,

thank you very much for your detailed explanations.
Unfortunately I'm not successful.
I have created a resource. When creating a new record I simply want a field to be filled out.
I tried your methods, unfortunately without success.

In Filament 2 I always do this with mount() { $this->form->fill([...])}Thank you very much for your detailed explanations.
Unfortunately I'm not successful.
I have created a resource. When creating a new record I simply want a field to be filled.
I tried your methods, unfortunately without success.

```
class CreateCar extends CreateRecord
{
    protected static string $resource = Car::class;
   
    protected function mutateFormDataBeforeFill(array $data): array
    {
        $data['type'] = 'e';

        return $data;
    }

}

```

In Filament 2 I always do this with mount() { $this->form->fill([...])}. This works fine!


Have you any further hint?

Thank you so much!

**Reply by leandrocfe** [reply]
If you are creating a record, you can use `afterFill` hook

```php
protected function afterFill(): void
{
    $this->form->fill([
        'type' => 'e',
    ]);
}
```

https://filamentphp.com/docs/3.x/panels/resources/creating-records#lifecycle-hooks

**Reply by photojo** [reply]
Ok! Perfect! This works fine! 
Thanks a lot!
